server/src/test/java/org/elasticsearch/cluster/coordination/NodeJoinTests.java

```diff
@@ -0,0 +1,399 @@
+        assertFalse(clusterStateHasNode(node1));
+        joinNodeAndRun(new JoinRequest(node1, Optional.of(new Join(node1, node0, 2, 1, 1))));
+        assertTrue(isLocalNodeElectedMaster());
+        assertTrue(clusterStateHasNode(node1));
+    }
+
+    public void testJoinAccumulation() {
+        DiscoveryNode node0 = newNode(0, true);
+        DiscoveryNode node1 = newNode(1, true);
+        setupFakeMasterServiceAndCoordinator(1, initialState(false, node0, 1, 1,
+            new VotingConfiguration(Collections.singleton(node1.getId()))));
+        assertFalse(isLocalNodeElectedMaster());
+        SimpleFuture fut = joinNodeAsync(new JoinRequest(node0, Optional.of(new Join(node0, node0, 2, 1, 1))));
+        deterministicTaskQueue.runAllTasks(random());
+        assertFalse(fut.isDone());
+        assertFalse(isLocalNodeElectedMaster());
+        joinNodeAndRun(new JoinRequest(node1, Optional.of(new Join(node1, node0, 2, 1, 1))));
+        assertTrue(isLocalNodeElectedMaster());
+        assertTrue(clusterStateHasNode(node1));
+        FutureUtils.get(fut);
+    }
+
+    public void testJoinFollowerFails() {
+        DiscoveryNode node0 = newNode(0, true);
+        DiscoveryNode node1 = newNode(1, true);
+        setupFakeMasterServiceAndCoordinator(1, initialState(false, node0, 1, 1,
+            new VotingConfiguration(Collections.singleton(node0.getId()))));
+        coordinator.coordinationState.get().handleStartJoin(new StartJoinRequest(node1, 2));
+        synchronized (coordinator.mutex) {
+            coordinator.becomeFollower("test", node1);
+        }
+        assertFalse(isLocalNodeElectedMaster());
+        assertThat(expectThrows(CoordinationStateRejectedException.class,
+            () -> joinNodeAndRun(new JoinRequest(node1, Optional.of(new Join(node1, node0, 2, 1, 1))))).getMessage(),
+            containsString("join target is a follower"));
+        assertFalse(isLocalNodeElectedMaster());
+    }
+
+    public void testBecomeFollowerFailsPendingJoin() {
+        DiscoveryNode node0 = newNode(0, true);
+        DiscoveryNode node1 = newNode(1, true);
+        setupFakeMasterServiceAndCoordinator(1, initialState(false, node0, 1, 1,
+            new VotingConfiguration(Collections.singleton(node1.getId()))));
+        SimpleFuture fut = joinNodeAsync(new JoinRequest(node0, Optional.of(new Join(node0, node0, 2, 1, 1))));
+        deterministicTaskQueue.runAllTasks(random());
+        assertFalse(fut.isDone());
+        assertFalse(isLocalNodeElectedMaster());
+        synchronized (coordinator.mutex) {
+            coordinator.becomeFollower("test", node1);
+        }
+        assertFalse(isLocalNodeElectedMaster());
+        assertThat(expectThrows(CoordinationStateRejectedException.class,
+            () -> FutureUtils.get(fut)).getMessage(),
+            containsString("following another master"));
+        assertFalse(isLocalNodeElectedMaster());
+    }
+
+    public void testConcurrentJoining() {
```

Comment:
Is there a way we could test this without introducing full multithreading (and therefore possibly nondeterminism)?

Reply:
I think we should also have more of these concurrent tests. I expect the non-concurrent functionality of this particular test to later be tested by "LegislatorTests".